src/ol_infrastructure/components/aws/auto_scale_group.py

```diff
@@ -0,0 +1,345 @@
+    image_id: str
+    instance_profile_arn: Union[str, pulumi.Output[str]]
+    instance_type: InstanceTypes = InstanceTypes.burstable_medium
+    key_name: str = "oldevops"
+    security_groups: List[Union[SecurityGroup, pulumi.Output]]
+    tag_specifications: List[TagSpecification]
+    user_data: Optional[Union[str, pulumi.Output[str]]]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class OLAutoScaleGroupConfig(AWSBase):
+    """Configuration Object for defining configuration needed to create an autoscale group."""
+
+    asg_name: str
+    aws_config: AWSBase
+    desired_size: PositiveInt = PositiveInt(2)
+    health_check_grace_period: NonNegativeInt = NonNegativeInt(0)
+    health_check_type: str = "ELB"
+    max_size: PositiveInt = PositiveInt(2)
+    min_size: PositiveInt = PositiveInt(1)
+    vpc_zone_identifiers: pulumi.Output[str]
+
+    instance_refresh_checkpoint_delay: PositiveInt = PositiveInt(3600)
+    instance_refresh_checkpoint_percentages: List[NonNegativeInt] = []
+    instance_refresh_warmup: PositiveInt = PositiveInt(health_check_grace_period)
+    instance_refresh_min_healthy_percentage: NonNegativeInt = NonNegativeInt(50)
+    instance_refresh_strategy: str = "Rolling"
+    instance_refresh_triggers: List[str] = ["tags"]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @validator("instance_refresh_strategy")
+    def is_valid_strategy(cls: "OLAutoScaleGroupConfig", strategy: str) -> str:
+        if strategy != "Rolling":
+            raise ValueError("The only vaild instance refresh strategy is 'Rolling'")
+        return strategy
+
+    @validator("health_check_type")
+    def is_valid_healthcheck(
+        cls: "OLAutoScaleGroupConfig", health_check_type: str
+    ) -> str:
+        if health_check_type not in ["ELB", "EC2"]:
+            raise ValueError(
+                f"health_check_type: {health_check_type} is not valid. Only 'ELB' or 'EC2' are accepted."
+            )
+        return health_check_type
+
+
+class OLAutoScaling(pulumi.ComponentResource):
+    """Component to create an autoscaling group with defaults as well as managed associated resources"""
+
+    lb_dns_name: str
+
+    def __init__(
+        self,
+        asg_config: OLAutoScaleGroupConfig,
+        lt_config: OLLaunchTemplateConfig,
+        tg_config: OLTargetGroupConfig,
+        lb_config: OLLoadBalancerConfig,
+        opts: pulumi.ResourceOptions = None,
+    ):
+
+        super().__init__(
+            "ol:infrastructure:aws:auto_scale_group:OLAutoScaleGroup",
+            asg_config.asg_name,
+            None,
+            opts,
+        )
+
+        # Shared attributes
+        resource_options = pulumi.ResourceOptions(parent=self).merge(opts)  # type: ignore
+        resource_name_prefix = asg_config.asg_name + "-"
+
+        pulumi.error(msg=f"{lb_config.subnets.__type__}", resource=lb_config.subnets)
+        # Create Load Balancer
+        load_balancer = LoadBalancer(
```

Comment:
For each attribute that is getting created it's useful to assign it as e.g. `self.load_balancer`, etc. That way when you create an instance of the component resource it's easy to access the child resources and their relevant attributes.